File: backend/src/main/java/com/java/NBE4_5_1_7/domain/openAI/OpenAiService.java
package com.java.NBE4_5_1_7.domain.openAI;

import com.java.NBE4_5_1_7.domain.openAI.dto.ChatRequest;
import com.java.NBE4_5_1_7.domain.openAI.dto.ChatResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import java.util.Arrays;
import java.util.List;

@Service
public class OpenAiService {

    private final WebClient webClient;

    @Value("${openai.api.key}")
    private String apiKey;

    public OpenAiService(WebClient.Builder webClientBuilder) {
        this.webClient = webClientBuilder.baseUrl("https://api.openai.com/v1").build();
    }

    /**
     * CS 인터뷰 시작 시 초기 질문 생성 (꼬리질문 규칙 포함)
     */
    public String askStartInterview(String interviewType) {
        Message systemMessage;
        if ("CS".equalsIgnoreCase(interviewType)) {
            systemMessage = new Message("system",
                    "당신은 신입 개발자를 대상으로 CS 지식에 관한 기술 면접을 진행하는 면접관입니다. " +
                            "지원자의 기초 CS 지식과 논리적 사고를 평가할 수 있는 구체적이고 현실적인 질문을 생성하세요. " +
                            "한 주제에 대해서는 최초 질문에 대한 후속(꼬리) 질문을 지원자의 답변을 듣고 하나씩 순차적으로 생성하며, " +
                            "최대 3개까지만 허용됩니다. 만약 지원자의 답변이 충분하다면 꼬리 질문 없이 바로 새로운 주제로 전환해 주세요. " +
                            "이상하거나 부적절한 질문은 절대 하지 마십시오.");
        } else if ("프로젝트".equalsIgnoreCase(interviewType)) {
            systemMessage = new Message("system",
                    "당신은 신입 개발자를 대상으로 프로젝트 경험에 관한 기술 면접을 진행하는 면접관입니다. 면접자의 프로젝트 경험에 대해 면접을 진행합니다. " +
                            "경험했던 프로젝트 주제가 무엇였고, 해당 프로젝트에서 맡은 역할이 무엇 이였나요? 라고 질문하세요" +
                            "첫 질문은 간단 명료하게 질문하세요." + "이상하거나 부적절한 단어를 붙이지 마세요" + "대화 주제와 내용에 상관없는 질문은 절대 하지 마십시오."
                            );
        } else {
            systemMessage = new Message("system", "면접을 시작합니다. 구체적이고 현실적인 질문만 생성하세요.");
        }

        List<Message> messages = Arrays.asList(systemMessage);
        ChatRequest chatRequest = new ChatRequest("gpt-3.5-turbo", messages);

        ChatResponse chatResponse = webClient.post()
                .uri("/chat/completions")
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .bodyValue(chatRequest)
                .retrieve()
                .bodyToMono(ChatResponse.class)
                .block();

        if (chatResponse != null && !chatResponse.getChoices().isEmpty()) {
            return chatResponse.getChoices().get(0).getMessage().getContent();
        }
        return "죄송합니다. 초기 질문을 생성하는 중 오류가 발생했습니다.";
    }

    /**
     * CS 인터뷰 진행 중 후속(꼬리) 질문 생성 (답변에 따라 꼬리 질문을 하나씩 생성하며, 최대 3개까지 허용)
     */
    public String askNextInterview(String interviewType, String answer) {
        Message systemMessage;
        if ("CS".equalsIgnoreCase(interviewType)) {
            systemMessage = new Message("system",
                    "당신은 신입 개발자를 대상으로 CS 지식에 관한 기술 면접을 진행하는 면접관입니다. " +
                            "지금은 지원자의 답변을 바탕으로 후속 질문(꼬리 질문)을 하나 생성해 주세요. " +
                            "단, 최초 질문에 대해 생성되는 꼬리 질문은 총 3개를 넘지 않아야 하며, " +
                            "지원자의 답변이 충분할 경우에는 꼬리 질문 없이 바로 새로운 주제로 전환할 수 있습니다. " +
                            "각 꼬리 질문은 답변 내용을 심도 있게 검증할 수 있도록 구체적이고 현실적으로 작성해 주세요. " +
                            "이상하거나 엉뚱한 질문은 하지 마십시오.");
        } else if ("프로젝트".equalsIgnoreCase(interviewType)) {
            systemMessage = new Message("system",
                    "면접자의 답변을 참고하여, " +
                            "먼저 프로젝트에서 맡은 역할과 개발한 기능에 대해 설명한 내용 중, " +
                            "해당 기능을 개발하는 과정에서 겪은 어려움과 그 해결 방안에 대해 구체적으로 질문해 주세요. " +
                            "또한, 만약 새로운 기술이나 방법을 도입하여 사용하셨다면, 그 선택 이유에 대해서도 후속 질문으로 추가할 수 있도록 해주세요. " +
                            "단, 후속 질문은 총 3개를 넘지 않아야 합니다.");
        } else {
            systemMessage = new Message("system", "면접을 진행합니다. 구체적이고 현실적인 후속 질문만 생성하세요.");
        }

        Message userMessage = new Message("user", "지원자의 답변: " + answer);
        List<Message> messages = Arrays.asList(systemMessage, userMessage);
        ChatRequest chatRequest = new ChatRequest("gpt-3.5-turbo", messages);

        ChatResponse chatResponse = webClient.post()
                .uri("/chat/completions")
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .bodyValue(chatRequest)
                .retrieve()
                .bodyToMono(ChatResponse.class)
                .block();

        if (chatResponse != null && !chatResponse.getChoices().isEmpty()) {
            return chatResponse.getChoices().get(0).getMessage().getContent();
        }
        return "죄송합니다. 후속 질문을 생성하는 중 오류가 발생했습니다.";
    }

    public String evaluateInterview(List<Message> conversation) {
        StringBuilder transcript = new StringBuilder();
        for (Message msg : conversation) {
            transcript.append(msg.getRole()).append(": ").append(msg.getContent()).append("\n");
        }
        String evalPrompt = "아래는 면접 질문 및 답변 목록입니다:\n\n" + transcript.toString() +
                "\n위 대화 내용을 바탕으로, 각 질문별로 지원자의 답변에서 부족했던 부분과 보완해야 할 점을 구체적이고 현실적으로 분석하고, " +
                "모범 답변을 제시해주세요. 단, 분석과 모범 답변은 불필요하게 추상적이거나 이상한 내용이 없도록 작성해야 합니다. " +
                "특정 기술이나 용어를 제외하고 모든 답변은 한국어로 해야 합니다.";

        Message systemMessage = new Message("system", evalPrompt);
        List<Message> messages = Arrays.asList(systemMessage);
        ChatRequest chatRequest = new ChatRequest("gpt-3.5-turbo", messages);

        ChatResponse chatResponse = webClient.post()
                .uri("/chat/completions")
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .bodyValue(chatRequest)
                .retrieve()
                .bodyToMono(ChatResponse.class)
                .block();

        if (chatResponse != null && !chatResponse.getChoices().isEmpty()) {
            return chatResponse.getChoices().get(0).getMessage().getContent();
        }
        return "죄송합니다. 평가 결과를 생성하는 중 오류가 발생했습니다.";
    }
}

File: backend/src/main/java/com/java/NBE4_5_3_7/domain/openAI/OpenAiService.kt
package com.java.NBE4_5_3_7.domain.openAI

import com.java.NBE4_5_3_7.domain.openAI.dto.ChatRequest
import com.java.NBE4_5_3_7.domain.openAI.dto.ChatResponse
import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import java.util.*

@Service
class OpenAiService(
    webClientBuilder: WebClient.Builder,
) {
    private val webClient: WebClient = webClientBuilder.baseUrl("https://api.openai.com/v1").build()

    @Value("\${openai.api.key}")
    private val apiKey: String? = null

    fun askStartInterview(interviewType: String?): String {
        val systemMessage = when {
            "CS".equals(interviewType, ignoreCase = true) -> {
                Message(
                    "system",
                    "당신은 신입 개발자를 대상으로 CS 지식에 관한 기술 면접을 진행하는 면접관입니다. " +
                            "지원자의 기초 CS 지식과 논리적 사고를 평가할 수 있는 구체적이고 현실적인 질문을 생성하세요. " +
                            "한 주제에 대해서는 최초 질문에 대한 후속(꼬리) 질문을 지원자의 답변을 듣고 하나씩 순차적으로 생성하며, " +
                            "최대 3개까지만 허용됩니다. 만약 지원자의 답변이 충분하다면 꼬리 질문 없이 바로 새로운 주제로 전환해 주세요. " +
                            "이상하거나 부적절한 질문은 절대 하지 마십시오."
                )
            }
            "프로젝트".equals(interviewType, ignoreCase = true) -> {
                Message(
                    "system",
                    "당신은 신입 개발자를 대상으로 프로젝트 경험에 관한 기술 면접을 진행하는 면접관입니다. 면접자의 프로젝트 경험에 대해 면접을 진행합니다. " +
                            "경험했던 프로젝트 주제가 무엇이었고, 해당 프로젝트에서 맡은 역할이 무엇이었나요? 라고 질문하세요. " +
                            "첫 질문은 간단 명료하게 하며, 이상하거나 부적절한 단어를 사용하지 말고, 대화 주제와 상관없는 질문은 하지 마십시오."
                )
            }
            else -> {
                Message("system", "면접을 시작합니다. 구체적이고 현실적인 질문만 생성하세요.")
            }
        }

        val chatRequest = ChatRequest("gpt-3.5-turbo", listOf(systemMessage))
        val chatResponse = sendToOpenAI(chatRequest)

        return extractContent(chatResponse, "초기 질문을 생성하는 중 오류가 발생했습니다.")
    }

    fun askNextInterview(interviewType: String?, answer: String?): String {
        if (answer == null) {
            return "답변이 제공되지 않았습니다. 다시 시도해주세요."
        }

        val systemMessage = when {
            "CS".equals(interviewType, ignoreCase = true) -> {
                Message(
                    "system",
                    "당신은 신입 개발자를 대상으로 CS 지식에 관한 기술 면접을 진행하는 면접관입니다. " +
                            "지금은 지원자의 답변을 바탕으로 후속 질문(꼬리 질문)을 하나 생성해 주세요. " +
                            "최대 3개까지만 허용하며, 답변이 충분하면 새로운 주제로 전환해도 됩니다. " +
                            "질문은 구체적이고 현실적이어야 하며, 이상하거나 엉뚱한 질문은 하지 마십시오."
                )
            }
            "프로젝트".equals(interviewType, ignoreCase = true) -> {
                Message(
                    "system",
                    "면접자의 프로젝트 관련 답변을 참고하여 후속 질문을 하나 생성하세요. " +
                            "해당 기능 구현 중 겪은 어려움과 해결 방안 등을 묻는 구체적 질문이어야 합니다. " +
                            "질문은 3개를 넘지 않아야 하며, 불필요하거나 엉뚱한 질문은 하지 마십시오."
                )
            }
            else -> {
                Message("system", "면접을 진행합니다. 구체적이고 현실적인 후속 질문만 생성하세요.")
            }
        }

        val userMessage = Message(
            "user",
            "지원자의 답변: $answer"
        )
        val chatRequest = ChatRequest("gpt-3.5-turbo", listOf(systemMessage, userMessage))
        val chatResponse = sendToOpenAI(chatRequest)

        return extractContent(chatResponse, "후속 질문을 생성하는 중 오류가 발생했습니다.")
    }

    fun evaluateInterview(conversation: List<Message>?): String {
        if (conversation.isNullOrEmpty()) {
            return "평가할 대화 내용이 없습니다."
        }

        val transcript = StringBuilder()
        for (msg in conversation) {
            transcript.append(msg.role ?: "unknown").append(": ").append(msg.content ?: "").append("\n")
        }

        val prompt = """
            아래는 면접 질문 및 답변 목록입니다:
            
            $transcript
            이 내용을 기반으로 질문별로 부족한 점, 보완할 점, 모범답안을 한국어로 구체적이고 현실적으로 작성해 주세요. 불필요하게 추상적이거나 이상한 표현은 배제해야 합니다.
            """.trimIndent()

        val chatRequest = ChatRequest("gpt-3.5-turbo", listOf(Message("system", prompt)))
        val chatResponse = sendToOpenAI(chatRequest)

        return extractContent(chatResponse, "평가 결과를 생성하는 중 오류가 발생했습니다.")
    }

    @CircuitBreaker(name = "openai", fallbackMethod = "fallbackOpenAi")
    fun sendToOpenAI(chatRequest: ChatRequest): ChatResponse? {
        return webClient.post()
            .uri("/chat/completions")
            .header("Authorization", "Bearer $apiKey")
            .header("Content-Type", "application/json")
            .bodyValue(chatRequest)
            .retrieve()
            .bodyToMono(ChatResponse::class.java)
            .block() // 동기 호출
    }

    fun fallbackOpenAi(chatRequest: ChatRequest, t: Throwable): ChatResponse {
        val fallbackMessage = Message("system", "현재 OpenAI 호출이 불가능합니다. 잠시 후 다시 시도해주세요.")

        val fallbackChoice = ChatResponse.Choice().apply {
            message = fallbackMessage
        }

        return ChatResponse().apply {
            choices = listOf(fallbackChoice)
        }
    }



//    private fun sendToOpenAI(chatRequest: ChatRequest): ChatResponse? {
//        return try {
//            webClient.post()
//                .uri("/chat/completions")
//                .header("Authorization", "Bearer $apiKey")
//                .header("Content-Type", "application/json")
//                .bodyValue(chatRequest)
//                .retrieve()
//                .bodyToMono(ChatResponse::class.java)
//                .block()
//        } catch (e: Exception) {
//            null
//        }
//    }

    private fun extractContent(chatResponse: ChatResponse?, fallback: String): String {
        return if (chatResponse != null && !chatResponse.choices.isNullOrEmpty()) {
            chatResponse.choices?.get(0)?.message?.content ?: "응답을 받지 못했습니다."
        } else {
            "죄송합니다. $fallback"
        }
    }
}